L1Trigger/TrackFindingTracklet/src/KFin.cc

```diff
@@ -0,0 +1,418 @@
+      for (int frame = 0; frame < (int)streamTrack.size(); frame++) {
+        const TTTrackRef& ttTrackRef = streamTrack[frame].first;
+        if (ttTrackRef.isNull()) {
+          input.push_back(nullptr);
+          continue;
+        }
+        //convert track parameter
+        const double r2Inv = digi(-ttTrackRef->rInv() / 2., baseUinv2R_);
+        const double phi0U =
+            digi(tt::deltaPhi(ttTrackRef->phi() - region_ * setup_->baseRegion() + setup_->hybridRangePhi() / 2.),
+                 baseUphiT_);
+        const double phi0S = digi(phi0U - setup_->hybridRangePhi() / 2., baseUphiT_);
+        const double cot = digi(ttTrackRef->tanL(), baseUcot_);
+        const double z0 = digi(ttTrackRef->z0(), baseUzT_);
+        const double phiT = digi(phi0S + r2Inv * digi(dataFormats_->chosenRofPhi(), baseUr_), baseUphiT_);
+        const double zT = digi(z0 + cot * digi(setup_->chosenRofZ(), baseUr_), baseUzT_);
+        // kill tracks outside of fiducial range
+        if (abs(phiT) > setup_->baseRegion() / 2. || abs(zT) > setup_->hybridMaxCot() * setup_->chosenRofZ() ||
+            abs(z0) > setup_->beamWindowZ()) {
+          input.push_back(nullptr);
+          continue;
+        }
+        // convert stubs
+        vector<Stub*> stubs;
+        stubs.reserve(channelAssignment_->numProjectionLayers(channel) + channelAssignment_->numSeedingLayers());
+        for (int layer = 0; layer < channelAssignment_->numProjectionLayers(channel); layer++) {
+          const FrameStub& frameStub = streamsStub[offsetStub + layer][frame];
+          const TTStubRef& ttStubRef = frameStub.first;
+          const int layerId = channelAssignment_->layerId(channel, layer);
+          if (ttStubRef.isNull())
+            continue;
+          // parse residuals from tt::Frame and take r and layerId from tt::TTStubRef
+          const bool barrel = setup_->barrel(ttStubRef);
+          const int layerIdTracklet = channelAssignment_->trackletLayerId(ttStubRef);
+          const double basePhi = barrel ? settings_->kphi1() : settings_->kphi(layerIdTracklet);
+          const double baseRZ = barrel ? settings_->kz(layerIdTracklet) : settings_->kz();
+          const int widthRZ = barrel ? setup_->tbWidthZ() : setup_->tbWidthR();
+          TTBV hw(frameStub.second);
+          const TTBV hwRZ(hw, widthRZ, 0, true);
+          hw >>= widthRZ;
+          const TTBV hwPhi(hw, setup_->tbWidthPhi(), 0, true);
+          hw >>= setup_->tbWidthPhi();
+          const double r = digi(setup_->stubR(hw, ttStubRef) - dataFormats_->chosenRofPhi(), baseUr_);
+          double phi = hwPhi.val(basePhi);
+          if (basePhi > baseUphi_)
+            phi += baseUphi_ / 2.;
+          const double z = digi(hwRZ.val(baseRZ) * (barrel ? 1. : -cot), baseUz_);
+          // determine module type
+          double rz = r + digi(dataFormats_->chosenRofPhi(), baseUr_);
+          if (barrel)
+            rz = rz * cot + z0 + z;
```

Comment:
I'm confused. Does "rz" represent z in barrel and r in endcap? Or something else? The eqn. "rz = rz * cot + ..." suggests it changes definition.

Reply:
I rewrote that a little